### PR TITLE
nodeAffinity filtered nodes should be excluded when calculating skew …

### LINKED ISSUE #106971: Potential bug of PodTopologySpread when nodeAffinity is specified
We know that nodeAffinity/nodeSelector is honored during the calculating of PodTopolgySpread, however, when some existing pods match the incoming pod's topologySpreadConstratints, while also excluded by the nodeAfffinity, things get a bit tricky. Raising this issue to discuss if we should read it as a bug.

Here are the repro steps:

1. Suppose you have a 3-node cluster, worker1 and worker2 belong to zoneA, and worker3 belongs to zoneB:
    ```
    kind-worker          Ready    kubernetes.io/hostname=kind-worker,zone=zoneA
    kind-worker2         Ready    kubernetes.io/hostname=kind-worker2,zone=zoneA
    kind-worker3         Ready    kubernetes.io/hostname=kind-worker3,zone=zoneB
    ```
2. And make a deployment with 3 replicas. Suppose 2 land on worker2, and 1 land on worker.
    ```
    pause-58dffb5c6b-4kb5z   1/1     Running   kind-worker2
    pause-58dffb5c6b-d4nzr   1/1     Running   kind-worker2
    pause-58dffb5c6b-jpv9t   1/1     Running   kind-worker3
    ```
3. Create a new deployment with topologySpreadConstraints as well as nodeAffinity (not schedule to worker2):
    ```yaml
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: test
    spec:
      replicas: 1
      selector:
        matchLabels:
          app: pause
      template:
        metadata:
          labels:
            app: pause
        spec:
          topologySpreadConstraints:
          - maxSkew: 1
            topologyKey: zone
            whenUnsatisfiable: DoNotSchedule
            labelSelector:
              matchLabels:
                app: pause
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                - matchExpressions:
                  - key: kubernetes.io/hostname
                    operator: NotIn
                    values: ["kind-worker2"]
          containers:
          - name: pause
            image: k8s.gcr.io/pause:3.6
    ```
4. The test pod will land on **worker3**.
    ```
    pause-58dffb5c6b-4kb5z   1/1     Running   kind-worker2
    pause-58dffb5c6b-d4nzr   1/1     Running   kind-worker2
    pause-58dffb5c6b-jpv9t   1/1     Running   kind-worker3
    test-7f87db4875-7gqxh    1/1     Running   kind-worker3
    ```

This behavior concerns me: it looks like the 2 pods on worker2 are taken into consideration, even though worker2 is excluded by nodeAffinity, so the pods on zoneA/zoneB distribute as **2:2**. However, shouldn't we disregard the matching pods that are excluded by nodeAffinity? In that case, the test pod should land on **worker1** to reach a pods distribution as **1:1**.

Please share your thoughts.

/sig scheduling
/kind bug

### PR DESCRIPTION
**Pull request by kerthcet**
…in PodTopologySpread

Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Fixes #106971

#### Does this PR introduce a user-facing change?

```release-note
ACTION REQUIRED: The calculations for Pod topology spread skew now excludes nodes that
don't match the node affinity/selector. This may lead to unschedulable pods if you previously had pods
matching the spreading selector on those excluded nodes (not matching the node affinity/selector),
especially when the topologyKey is not node-level. Revisit the node affinity and/or pod selector in the
topology spread constraints to avoid this scenario.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by alculquicondor — COMMENTED** [review]
This change seems to have no effect. Where are the tests?

**Comment by alculquicondor** [comment]
However, we need to decide if this is worth fixing https://github.com/kubernetes/kubernetes/issues/106971#issuecomment-993604231

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 1]
```diff
@@ -222,6 +222,8 @@ func (pl *PodTopologySpread) calPreFilterState(pod *v1.Pod) (*preFilterState, er
 		TpKeyToCriticalPaths: make(map[string]*criticalPaths, len(constraints)),
 		TpPairToMatchNum:     make(map[topologyPair]*int32, sizeHeuristic(len(allNodes), constraints)),
 	}
+
+	var availableNodes []*framework.NodeInfo
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:226]: availableNodes -> filteredNodes.

And add a comment above:

```
// Nodes that pass nodeAffinity check and carry required topology keys will be
// stored in `filteredNodes`, and be looped later to calculate preFilterState.
```

**Review by Huang-Wei — COMMENTED** [review]
Can you add a UT like this:

```go
		{
			name: "incoming pod has nodeAffinity, pods spread as 0/~2~/1/0, hence node-a fits",
			pod: st.MakePod().Name("p").Label("foo", "").
				NodeAffinityNotIn("node", []string{"node-b"}).
				SpreadConstraint(1, "zone", v1.DoNotSchedule, st.MakeLabelSelector().Exists("foo").Obj()).
				Obj(),
			nodes: []*v1.Node{
				st.MakeNode().Name("node-a").Label("zone", "zone1").Label("node", "node-a").Obj(),
				st.MakeNode().Name("node-b").Label("zone", "zone1").Label("node", "node-b").Obj(),
				st.MakeNode().Name("node-x").Label("zone", "zone2").Label("node", "node-x").Obj(),
				st.MakeNode().Name("node-y").Label("zone", "zone2").Label("node", "node-y").Obj(),
			},
			existingPods: []*v1.Pod{
				st.MakePod().Name("p-b1").Node("node-b").Label("foo", "").Obj(),
				st.MakePod().Name("p-b2").Node("node-b").Label("foo", "").Obj(),
				st.MakePod().Name("p-y1").Node("node-y").Label("foo", "").Obj(),
			},
			wantStatusCode: map[string]framework.Code{
				"node-a": framework.Success,
				"node-b": framework.Success, // in real case, it's Unschedulable
				"node-x": framework.Unschedulable,
				"node-y": framework.Unschedulable,
			},
		},
```

**Comment by Huang-Wei** [comment]
> This change seems to have no effect.

@alculquicondor it has. See the original issue details, or the UT I commented.

**Comment by alculquicondor** [comment]
Ah, right, I read the PR too fast.

Is there any scenario where this change would cause pods to be unschedulable compared to the previous calculation?

**Comment by kerthcet** [comment]
> Ah, right, I read the PR too fast.
> 
> Is there any scenario where this change would cause pods to be unschedulable compared to the previous calculation?

AFAIK, we just filtered out unsatisfied nodes, these nodes will never be scheduled to, so I think it has no side effects.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go (hunk 1 above)
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: please take a look.

**Comment by alculquicondor** [comment]
We are changing the calculation of the skew, so it has effects.

Let's say we have nodes A1, A2 (in zone A), B1 and B2 (in zone B), each with one matching pod. The incoming pod has anti-affinity to node B2. The spreading constraint is over zone, with maxSkew=1.
In the old behavior, the matching pods for the zones are {2, 2}, so the pod could go to nodes A1, A2, B1.
In the new behavior, the matching pods for the zones are {2, 1}, so the pod could only go to node B1.

If B1 is full, but A1 and A2 is not, this pod is now unschedulable, but it was schedulable before. Thus, this is a breaking change.

**Comment by Huang-Wei** [comment]
> If B1 is full, but A1 and A2 is not, this pod is now unschedulable, but it was schedulable before. Thus, this is a breaking change.

@alculquicondor I think we're fixing a bug :) instead of behavior change. I'd expect this PR to be back ported.

**Comment by alculquicondor** [comment]
At the very least the semantics are unclear. The point is that the change could lead to unschedulable pods and that could break production workloads. At least in a minor release, the best practice is to have a canary cluster, so maybe we could limit the change to 1.24. That could still cause problems, but maybe less.

**Comment by alculquicondor** [comment]
But otherwise, I'm not convinced that we are serving a valid use case and I want to minimize the risk.

**Comment by Huang-Wei** [comment]
I don't think the semantics is unclear. Here is a similar example: podA lands on a node that carries a label `env=qa`, and afterwards the node's label gets cleared. In this case, should we consider upcoming pods to consider podA? (suppose the required topology is `env`) The current behavior is no b/c the topology podA lands on doesn't have the required topologyKey. The same rationale should be applied to unqualified nodes (failed on nodeAffinity/nodeSelector) as well.

The use case might be a bit hypothetical, but we need to fix unclear/inconsistent semantics.

> that could break production workloads.

If the workloads used to run in an incorrect way, we should correct it instead of keeping them as-is.

**Comment by alculquicondor** [comment]
@liggitt for thoughts

**Comment by 2014BDuck** [comment]
Hi, May I share my point of view here? I think currently the behavior is not meet my expectation so it should be considered as a bug.

However, there might be k8s users who use this combination: `topologySpreadConstraints` + `nodeAffinity` and config accordingly. With this change, there might be different schedule results which out of **their expectation**?

So for those who use those features (wrongly, tho), any idea to minimize the changes? Or do we simply consider it as a bug and tell users to change what they've configured?

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 2]
```diff
@@ -1400,6 +1400,32 @@ func TestSingleConstraint(t *testing.T) {
 				"node-b": framework.Unschedulable,
 			},
 		},
+		{
+			// In this unit test, we don't run NodeAffinity Predicate, so node-b
+			// still expected to be fits;
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1405]: In a unit test, NodeAffinity plugin is not involved, so node-b still fits.

**Review by Huang-Wei — COMMENTED** [review]
Just a nit.

**Comment by Huang-Wei** [comment]
This PR looks good to me. I will approve it if @liggitt or @alculquicondor, based on a lazy consensus.

**Comment by alculquicondor** [comment]
I don't think the comments I've left qualify for lazy consensus. I'm still awaiting for a strong justification for the potential risk.

**Comment by alculquicondor** [comment]
Something that is very clear to me is that we shouldn't backport this. Whether we should apply the change is still on debate.

**Comment by Huang-Wei** [comment]
The decision depends on how we interpret the nodeAffinity a workload carries along with topologySpreadConstraints: I'm inclined to read it as "well, this pod **explicitly** specified nodeAffinity, which means it knows where to schedule, i.e., it demands to spread its pods among the **targeting** nodes, without considering the pods in the _excluded_ nodes".

> I'm still awaiting for a strong justification for the potential risk.

The risk is fairly low IMO as the trigger conditions are:

1. there are matching pods on excluded node (say the incoming workload has both nodeAffinity and topologySpreadConstraints specified)
2. and, the topology key specified in the topologySpreadConstraint is not node-level.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go (hunk 2 above)
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1405]: done.

**Comment by kerthcet** [comment]
IMO, how we treat this issue depends on the original design and whether end-uses understood the intension.  

The original design let's refer to this document: [Interaction With Node Affinity and Node Selectors](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#interaction-with-node-affinity-and-node-selectors), the scheduler will skip the non-matching nodes from the skew calculations if the incoming Pod has spec.nodeSelector or spec.affinity.nodeAffinity defined. Depending on this, I treat this issue as a bug.

How end-users understood this intension? I agree with @Huang-Wei , at least for me, I thought we would excluded these nodes unsatisfied with nodeAffinity/nodeLabel. 

Some poor thoughts.

**Comment by alculquicondor** [comment]
Based on https://github.com/kubernetes/kubernetes/pull/107009#issuecomment-1007006991, ok for 1.24+ only. Put an `ACTION REQUIRED` in the release notes.

**Comment by kerthcet** [comment]
> Based on [#107009 (comment)](https://github.com/kubernetes/kubernetes/pull/107009#issuecomment-1007006991), ok for 1.24+ only. Put an `ACTION REQUIRED` in the release notes.

done.

**Comment by kerthcet** [comment]
> You have to say what is the required action, along with what the change is about. Can you think of a short set of instructions for users?

please take a look.

**Comment by alculquicondor** [comment]
My suggestion:

```
ACTION REQUIRED: The calculations for Pod topology spread skew now excludes nodes that don't match the node affinity. This may lead to unschedulable pods if you previously had pods matching the spreading selector in nodes that don't match the node affinity. Revisit the node affinity and/or pod selector in the topology spread constraints to avoid this scenario.
```

**Comment by kerthcet** [comment]
> My suggestion:
> 
> ```
> ACTION REQUIRED: The calculations for Pod topology spread skew now excludes nodes that don't match the node affinity. This may lead to unschedulable pods if you previously had pods matching the spreading selector in nodes that don't match the node affinity. Revisit the node affinity and/or pod selector in the topology spread constraints to avoid this scenario.
> ```

thanks aldo, updated already.

**Comment by Huang-Wei** [comment]
Some minor updates based on @alculquicondor 's comment: (also make it multi-lined for readability)

```
ACTION REQUIRED: The calculations for Pod topology spread skew now excludes nodes that
don't match the node affinity/selector. This may lead to unschedulable pods if you previously had pods
matching the spreading selector on those excluded nodes (not matching the node affinity/selector),
especially when the topologyKey is not node-level. Revisit the node affinity and/or pod selector in the
topology spread constraints to avoid this scenario.
```

**Comment by kerthcet** [comment]
> Some minor updates based on @alculquicondor 's comment: (also make it multi-lined for readability)
> 
> ```
> ACTION REQUIRED: The calculations for Pod topology spread skew now excludes nodes that
> don't match the node affinity/selector. This may lead to unschedulable pods if you previously had pods
> matching the spreading selector on those excluded nodes (not matching the node affinity/selector),
> especially when the topologyKey is not node-level. Revisit the node affinity and/or pod selector in the
> topology spread constraints to avoid this scenario.
> ```

Thanks wei. Done.

**Pull request merged**